improver/psychrometric_calculations/hail_size.py

```diff
@@ -0,0 +1,338 @@
+                self.model_id_attr
+            ]
+
+        hail_size_cube = create_new_diagnostic_cube(
+            name="size_of_hail_stones",
+            units="mm",
+            template_cube=ccl_temperature,
+            data=hail_size,
+            mandatory_attributes=generate_mandatory_attributes(
+                [ccl_temperature, ccl_pressure]
+            ),
+            optional_attributes=attributes,
+        )
+
+        return hail_size_cube
+
+    def check_cubes(
+        self,
+        ccl_temperature,
+        ccl_pressure,
+        temperature_on_pressure,
+        humidity_mixing_ratio_on_pressure,
+    ):
+        """function to check size and units of input cubes"""
+
+        temp_slice = next(temperature_on_pressure.slices_over("pressure"))
+
+        assert_spatial_coords_match([ccl_temperature, ccl_pressure, temp_slice])
+        assert_spatial_coords_match(
+            [temperature_on_pressure, humidity_mixing_ratio_on_pressure]
+        )
+
+        ccl_temperature.convert_units("K")
+        ccl_pressure.convert_units("Pa")
+        temperature_on_pressure.convert_units("K")
+        humidity_mixing_ratio_on_pressure.convert_units("kg/kg")
+
+    def process(
+        self,
+        ccl_temperature: Cube,
+        ccl_pressure: Cube,
+        temperature_on_pressure: Cube,
+        humidity_mixing_ratio_on_pressure: Cube,
+    ) -> Cube:
+        """
+        Args:
+            ccl_temeprature:
+                cube of the cloud condensation level temperature
+            ccl_pressure:
+                cube of the cloud condensation level pressure.
+            temperature_on_pressure:
+                cube of temperature on pressure levels 
+            humidity_mixing_ratio_on_pressure:
+                cube of humidity mixing ratio on pressure levels
+        Returns:
+            Cube of hail size
+        """
+
+        self.check_cubes(
+            ccl_temperature,
+            ccl_pressure,
+            temperature_on_pressure,
+            humidity_mixing_ratio_on_pressure,
+        )
+
+        np.ma.masked_less(ccl_temperature.data, 268.15) # any points where this is true leads to negative vertical/horizontal values and so no hail size
+
+        pressure_at_268, temperature_at_268 = self.pressure_at_268(
+            temperature_on_pressure
+        )
+
+        humidity_mixing_ratio_at_268 = self.humidity_mixing_ratio_at_268(
+            humidity_mixing_ratio_on_pressure, pressure_at_268
+        )
+
+        B = temperature_at_268
+        c = self.dry_adiabatic_drop_to_ccl(
+            ccl_pressure, temperature_at_268, pressure_at_268
+        )
+        b = self.potential_temperature_after_saturated_ascent_from_CCL(
+            ccl_temperature,
+            ccl_pressure,
+            pressure_at_268,
+            humidity_mixing_ratio_at_268,
+            ccl_temperature,
+        )
+
+        horizontal = c - B
+        vertical = b - B
```

Comment:
I think it would be a good idea to refactor this section into a separate function where the doc-string can describe what `B, b, C, c` are. Without reference to the documentation, they're extremely opaque! It might be best to give them human-readable names.